test/modules/EdgeHubRestartTester/src/Program.cs

```diff
@@ -0,0 +1,363 @@
+                    else
+                    {
+                        // something is wrong, Log and re-throw
+                        Logger.LogError(e, $"[SendDirectMethodAsync] Exception caught with SequenceNumber {Interlocked.Read(ref directMethodCount).ToString()}");
+                        throw;
+                    }
+                }
+            }
+
+            return new Tuple<DateTime, HttpStatusCode>(DateTime.UtcNow, HttpStatusCode.InternalServerError);
+        }
+
+        static bool IsEdgeHubDownDuringDirectMethodSend(IotHubCommunicationException e)
+        {
+            // This is a socket exception error code when EdgeHub is down.
+            const int EdgeHubNotAvailableErrorCode = 111;
+
+            if (e?.InnerException?.InnerException is SocketException)
+            {
+                int errorCode = ((SocketException)e.InnerException.InnerException).ErrorCode;
+                return errorCode == EdgeHubNotAvailableErrorCode;
+            }
+
+            return false;
+        }
+
+        static async Task<Tuple<DateTime, HttpStatusCode>> SendMessageAsync(
+            ModuleClient moduleClient,
+            string trackingId,
+            Guid batchId,
+            string msgOutputEndpoint,
+            DateTime runExpirationTime,
+            CancellationToken cancellationToken)
+        {
+            while ((!cancellationToken.IsCancellationRequested) && (DateTime.UtcNow < runExpirationTime))
+            {
+                Message message = new Message(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(new { data = DateTime.UtcNow.ToString() })));
+                Interlocked.Increment(ref messageCount);
+                message.Properties.Add("sequenceNumber", Interlocked.Read(ref messageCount).ToString());
+                message.Properties.Add("batchId", batchId.ToString());
+                message.Properties.Add("trackingId", trackingId);
+
+                try
+                {
+                    // Sending the result via edgeHub
+                    await moduleClient.SendEventAsync(msgOutputEndpoint, message);
+                    Logger.LogInformation($"[SendMessageAsync] Send Message with count {Interlocked.Read(ref messageCount).ToString()}: finished.");
+                    return new Tuple<DateTime, HttpStatusCode>(DateTime.UtcNow, HttpStatusCode.OK);
+                }
+                catch (TimeoutException ex)
+                {
+                    // TimeoutException is expected to happen while the EdgeHub is down.
+                    // Let's log the attempt and retry the message send until successful
+                    Logger.LogDebug(ex, $"[SendMessageAsync] Exception caught with SequenceNumber {messageCount}, BatchId: {batchId.ToString()};");
+                }
+            }
+
+            return new Tuple<DateTime, HttpStatusCode>(DateTime.UtcNow, HttpStatusCode.InternalServerError);
+        }
+
+        static async Task<Tuple<DateTime, HttpStatusCode>> RestartModules(
```

Comment:
Consider moving these methods to a new class to keep Program as small as possible